op-e2e/e2eutils/disputegame/output_game_helper.go

```diff
@@ -459,6 +465,94 @@ func (g *OutputGameHelper) ResolveClaim(ctx context.Context, claimIdx int64) {
+				return claim.Attack(ctx, correct)
+			} else if execClaimPosition.TraceIndex(execDepth).Uint64() > targetTraceIndex {
+				g.t.Logf("Bisecting: Attack incorrectly for step")
+				return claim.Attack(ctx, common.Hash{0xdd})
+			} else if execClaimPosition.TraceIndex(execDepth).Uint64()+1 == targetTraceIndex {
+				g.t.Logf("Bisecting: Defend incorrectly for step")
+				return claim.Defend(ctx, common.Hash{0xcc})
+			} else {
+				newPosition := execClaimPosition.Defend()
+				correct, err := provider.Get(ctx, newPosition)
+				g.require.NoError(err)
+				g.t.Logf("Bisecting: Defend correctly for step at newPosition=%v execIndexAtDepth=%v", newPosition, newPosition.TraceIndex(execDepth))
+				return claim.Defend(ctx, correct)
+			}
+		}
+
+		// Attack or Defend depending on whether the claim we're responding to is to the left or right of the trace index
+		// Induce the honest challenger to attack or defend depending on whether our new position will be to the left or right of the trace index
+		if execClaimPosition.TraceIndex(execDepth).Uint64() < targetTraceIndex && claim.Depth() != splitDepth+1 {
+			newPosition := execClaimPosition.Defend()
+			if newPosition.TraceIndex(execDepth).Uint64() < targetTraceIndex {
+				g.t.Logf("Bisecting: Defend correct. newPosition=%v execIndexAtDepth=%v", newPosition, newPosition.TraceIndex(execDepth))
+				correct, err := provider.Get(ctx, newPosition)
+				g.require.NoError(err)
+				return claim.Defend(ctx, correct)
+			} else {
+				g.t.Logf("Bisecting: Defend incorrect. newPosition=%v execIndexAtDepth=%v", newPosition, newPosition.TraceIndex(execDepth))
+				return claim.Defend(ctx, common.Hash{0xaa})
+			}
+		} else {
+			newPosition := execClaimPosition.Attack()
+			if newPosition.TraceIndex(execDepth).Uint64() < targetTraceIndex {
+				g.t.Logf("Bisecting: Attack correct. newPosition=%v execIndexAtDepth=%v", newPosition, newPosition.TraceIndex(execDepth))
+				correct, err := provider.Get(ctx, newPosition)
+				g.require.NoError(err)
+				return claim.Attack(ctx, correct)
+			} else {
+				g.t.Logf("Bisecting: Attack incorrect. newPosition=%v execIndexAtDepth=%v", newPosition, newPosition.TraceIndex(execDepth))
+				return claim.Attack(ctx, common.Hash{0xbb})
+			}
+		}
+	}
+
+	g.LogGameData(ctx)
+	// Initial bisect to put us on defense
+	claim := bisectTraceIndex(outputRootClaim)
+	g.DefendClaim(ctx, claim, bisectTraceIndex)
+}
+
+func (g *OutputGameHelper) PreimageExistsInOracle(ctx context.Context, data *types.PreimageOracleData) bool {
+	caller := batching.NewMultiCaller(g.system.NodeClient("l1").Client(), batching.DefaultBatchSize)
+	contract, err := contracts.NewFaultDisputeGameContract(g.addr, caller)
+	g.require.NoError(err, "Failed to create game contract")
+	vm, err := contract.Vm(ctx)
+	g.require.NoError(err, "Failed to create vm contract")
+	oracle, err := vm.Oracle(ctx)
+	g.require.NoError(err, "Failed to create oracle contract")
```

Comment:
You can just use the `GetOracle` method on fault dispute game contract. Then we can skip making the Vm method public.

Reply:
`GetOracle` returns a `IPreimageOracle` interface, which doesn't (yet) expose methods to read the `preimagePart`.
@clabby We should consider adding existence checks to the `IPreimageOracle.sol` interface to facilitate this.

Reply:
I don't think it's actually tied to `IPreimageOracle` - it just returns `*PreimageOracleContract`.  It's a convenient helper, not a direct mapping to contract calls: https://github.com/ethereum-optimism/optimism/blob/f64801693459469df5ab6d8012adf73f1673673c/op-challenger/game/fault/contracts/faultdisputegame.go#L128-L134

Reply:
gotcha. I misread that as the binding contract.